### PR TITLE
cli: Add 'agent generate-config' sub-command

### PR DESCRIPTION
**Pull request by averche**
> Note: documentation will be added in a different PR

## Background

This pull request implements a new sub-command `agent generate-config`. It is part of the larger effort to add environment variable support within Vault Agent ([VLT-253](https://docs.google.com/document/d/1HLTkPtxUTtW-wMkePWoHmZoGt9lLg2POslt1ranaWE8/edit#)).

The goal of this sub-command is to help new users with an easy on-boarding to Vault, specifically for integrating Vault into their applications. It will recurse through the Vault path tree for the given path(s) and produce an `agent.hcl` configuration file with `token_file` authentication method and default environment-variable-to-secret mappings. The file can then be used by Vault Agent in a process supervisor mode (will be implemented in future PRs).

For the first release of this template configuration helper, we will only support KV-V1 and KV-V2 secret types.

## Example

```sh
$ vault agent generate-config \
       -type="env-template" \
       -exec="./my-app.sh" \
       -path="secret/test" \
       -path="secret/my-app/*"                                                           

Successfully generated "agent.hcl" configuration file!
```

agent.hcl:

```hcl
auto_auth {

  method {
    type = "token_file"

    config {
      token_file_path = "/Users/avean/.vault-token"
    }
  }
}

template_config {
  static_secret_render_interval = "30s"
  exit_on_retry_failure         = true
}

vault {
  address = "http://localhost:8200"
}

env_template "TEST_PASSWORD" {
  contents             = "{{ with secret \"secret/data/test\" }}{{ .Data.data.password }}{{ end }}"
  error_on_missing_key = true
}
env_template "TEST_USER" {
  contents             = "{{ with secret \"secret/data/test\" }}{{ .Data.data.user }}{{ end }}"
  error_on_missing_key = true
}
env_template "API_KEY_VAL1" {
  contents             = "{{ with secret \"secret/data/my-app/api-key\" }}{{ .Data.data.val1 }}{{ end }}"
  error_on_missing_key = true
}
env_template "DATABASE_PASSWORD" {
  contents             = "{{ with secret \"secret/data/my-app/database\" }}{{ .Data.data.password }}{{ end }}"
  error_on_missing_key = true
}
env_template "DATABASE_USER" {
  contents             = "{{ with secret \"secret/data/my-app/database\" }}{{ .Data.data.user }}{{ end }}"
  error_on_missing_key = true
}
env_template "BAR_VAL1" {
  contents             = "{{ with secret \"secret/data/my-app/nested/bar\" }}{{ .Data.data.val1 }}{{ end }}"
  error_on_missing_key = true
}
env_template "BAR_VAL2" {
  contents             = "{{ with secret \"secret/data/my-app/nested/bar\" }}{{ .Data.data.val2 }}{{ end }}"
  error_on_missing_key = true
}
env_template "FOO_VAL1" {
  contents             = "{{ with secret \"secret/data/my-app/nested/foo\" }}{{ .Data.data.val1 }}{{ end }}"
  error_on_missing_key = true
}
env_template "FOO_VAL2" {
  contents             = "{{ with secret \"secret/data/my-app/nested/foo\" }}{{ .Data.data.val2 }}{{ end }}"
  error_on_missing_key = true
}

exec {
  command                   = ["./my-app.sh"]
  restart_on_secret_changes = "always"
  restart_stop_signal       = "SIGTERM"
}

```

## Related pull requests

- https://github.com/hashicorp/vault/pull/20464
- https://github.com/hashicorp/vault/pull/20488

[VLT-253]: https://hashicorp.atlassian.net/browse/VLT-253?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: command/agent_generate_config.go
Comment (dhuckins) [inline command/agent_generate_config.go:139]: should we look for env var `VAULT_TOKEN` here too?

File: command/agent_generate_config.go
Comment (dhuckins) [inline command/agent_generate_config.go:151]: this might be a little quick, think something like 10 minutes is the default

File: command/agent_generate_config.go
Comment (dhuckins) [inline command/agent_generate_config.go:170]: should the default be stdout?

File: command/agent_generate_config.go
Comment (VioletHynes) [inline command/agent_generate_config.go:58]: Even though we only have one option today, I think it might make sense for there to be no default for this value.

Today, the main mental model for Agent templating is that it generates files, so an `agent generate-config` option is something that a lot of customers might think is targeting that use case. Making the user specify `agent generate-config -type=env-template` makes it clear that it's creating env-template config, and means that if we do add generate-config for file templating, the default won't be confusing or make it seem like we're favouring one approach over another.

Comment (averche) [reply command/agent_generate_config.go:58]: Good point, thanks!

File: command/agent_generate_config.go
Comment (VioletHynes) [inline command/agent_generate_config.go:135]: Is it possible to generate a comment as part of the HCL with this HCL library?

If so, I think it would make sense to leave a comment warning that we'd recommend another auto auth method for production (perhaps with a link to the auto-auth docs).

File: command/agent_generate_config_test.go
Comment (VioletHynes) [inline command/agent_generate_config_test.go:27]: Why do we have a time.Sleep here? Can we remove it?

File: command/agent_generate_config_test.go
Comment (VioletHynes) [inline command/agent_generate_config_test.go:181]: From what I can tell, we only seem to be testing the `env_template` stanzas thoroughly in this test. I'd like to see either a different test or an enhancement to this test where we test the generation of the auto auth and exec stanzas, too.

File: command/agent_generate_config_test.go
Comment (dhuckins) [inline command/agent_generate_config_test.go:16]: after we add the config, should add tests that the generated config should be readable

**Comment by VioletHynes** [comment]
Looks great! Added some comments so far but broadly it looks fantastic. One thing I'm also interested in: we should add docs for this subcommand. I don't really mind if that's as part of a different PR or if it's part of this one, but we should make sure docs are added. I'm happy to review them all the same.

File: command/agent_generate_config.go
Comment (averche) [reply command/agent_generate_config.go:144]: It turned out to be a lot more difficult than I imagined 😄 The `gohcl.EncodeIntoBody(...)` which takes a struct and marshals it into HCL does not currently support adding comments. The recommendation from the maintainers of the library was to use the lower level `hclwrite` instead. This would work but we'd need to manually recreate [this function](https://github.com/hashicorp/hcl/blob/v2.16.2/gohcl/encode.go#L88-L194) and add the comment logic. Perhaps the right approach would be to add support for this in the library. I'd say this is a phase 2 feature :)

File: command/agent_generate_config_test.go
Comment (averche) [reply command/agent_generate_config_test.go:181]: This is probably the most complex part of the code so I wanted to test it thoroughly. I'll add another test for the overall config 👍

File: command/agent_generate_config.go
Comment (averche) [reply command/agent_generate_config.go:139]: I don't think `VAULT_TOKEN` is currently supported by agent.

File: command/agent_generate_config.go
Comment (averche) [reply command/agent_generate_config.go:144]: Oooh that's a great idea, thanks, will add it in the output!

File: command/agent_generate_config_test.go
Comment (averche) [reply command/agent_generate_config_test.go:27]:
```suggestion
```
Nice, let's remove it then!

File: command/agent_generate_config_test.go
Comment (averche) [reply command/agent_generate_config_test.go:16]: Good call, we definitely should

File: command/agent_generate_config_test.go
Comment (averche) [inline command/agent_generate_config_test.go:35]:
```suggestion
```

File: command/agent_generate_config.go
Comment (dhuckins) [inline command/agent_generate_config.go:362]:
```suggestion
	StaticSecretRenderInterval string `hcl:"static_secret_render_interval"`
```

File: command/agent_generate_config.go
Comment (averche) [inline command/agent_generate_config.go:151]:
```suggestion
			StaticSecretRendereInterval: "5m",
```

Thanks! Updating

**Comment by averche** [comment]
> Looks great! Added some comments so far but broadly it looks fantastic. One thing I'm also interested in: we should add docs for this subcommand. I don't really mind if that's as part of a different PR or if it's part of this one, but we should make sure docs are added. I'm happy to review them all the same.

Thanks! Yes, docs is definitely next 👍 I will probably hold off on merging this till the docs PR is ready.

File: command/agent_generate_config.go
Comment (averche) [reply command/agent_generate_config.go:170]: Maybe 🤷 We kind of want to give a warning about the token auto-auth method so I'm slightly leaning towards writing to a file. We could also require a file to be specified.

File: command/agent_generate_config.go
Comment (averche) [reply command/agent_generate_config.go]: Rendere sounds almost french  😄

**Comment by VioletHynes** [comment]
Just to make sure we're not waiting on each other: I'm hoping to see the note about changing the auto_auth type and the additional tests for the other parts of generated config before I approve.

Once these are added, feel free to re-request me as a viewer to ping me!

File: command/agent_generate_config.go
Comment (averche) [reply command/agent_generate_config.go:178]: Added a warning in the output ([5a9e550](https://github.com/hashicorp/vault/pull/20530/commits/5a9e550fddf3bf41c36c585d5a75c6ff08cb6e84)), thanks again for the suggestion!

File: command/agent_generate_config_test.go
Comment (averche) [reply command/agent_generate_config_test.go:143]: Refactored a bit and added a couple more tests in [1f83221](https://github.com/hashicorp/vault/pull/20530/commits/1f83221aacff779fbd68c9d8f1bf7f185f13ab33)

**Comment by averche** [comment]
> Just to make sure we're not waiting on each other: I'm hoping to see the note about changing the auto_auth type and the additional tests for the other parts of generated config before I approve.
> 
> Once these are added, feel free to re-request me as a viewer to ping me!

Thanks! I added both

**Review by VioletHynes — APPROVED** [review]
Looks great! Nice work.

**Review by dhuckins — APPROVED** [review]

File: changelog/20530.txt
Comment (dhuckins) [reply changelog/20530.txt:1]: @mladlow should we also make that change for https://github.com/hashicorp/vault/pull/20739 ?

Comment (mladlow) [reply changelog/20530.txt:1]: @dhuckins Are these two PRs the same or different "features"? If you scroll through historical changelog sections for major releases, you can see what's been called out in the "features" section. For the features section, it's expected that features will have multiple PRs and it doesn't always make sense to have a changelog entry for every PR that composes a major new feature.

Comment (dhuckins) [reply changelog/20530.txt:1]: gotcha, same feature (at least they are related). so just one should be fine. thanks!

**Pull request merged**
